### ISSUE TITLE
Support for multiple auth endpoints?

### ISSUE THREAD

**Issue by xviakoh** [pusher/pusher-js#83]
Hi, 

I'm wondering if there's a plan for supporting different auth endpoint per channel. Current client implementation looks like it's grabbing auth endpoint from static global config variable which is bit limiting.

In our use case, we want to be able to provide different authentication end point for different channels so that certain channels are only available for certain types of users. For example, there are two private channels. One is available for all logged in users. The other is only available if user has certain permission. For first channel, authentication endpoint will simply check if user is logged into our site. For later channel, auth endpoint needs to check user's permission in details.

We've tried modifying global config's authEndpoint variable on the fly just before calling subscribe() but this approach is prone to race condition. If multiple subscribe() methods are called within various of different points in our javascript app, it's not concurrency safe since the variable is static.

I've started looking into js code and it doesn't seem too bad of idea if subscribe() allowed taking optional parameter for auth end point url, pass it all the way down to channel obj, and eventually down to authorizer obj.

Do you see any problem having multiple endpoints architecturally? Am I missing any unknown issues with this approach?

Thanks,
Koh

**Comment by sandstrom** [comment]
If the pusher js-client could take a function `authAdapter` returning a promise, then it could be left to the user to perform the authentication in any way fit (different endpoints, headers handling, useCredentials switch, etc).

```js
new Pusher(apiKey, {
  authAdapter: function(channel, ...) {
    return Promise(resolve, reject) {
      // ...
    };
  },
  // other conf...
});
```

**Comment by joeenzminger** [comment]
This would be very useful - being able to specify the authEndpoint per subscription + being able to create an authAdapter as suggested by @sandstrom makes the client library completely agnostic to the authentication implementation details.  I'd rather not have to keep (and look up) auth endpoints based on channel names.  In our app, we use an entities' ID as it's channel name (prefixed with private- or presence-, and each entity has an href (that may or may not include the ID).  What I'd love to be able to do is something like:

`pusher.subscribe('presence-' + entity.id, entity.href+'/Authorize')`

**Comment by sandstrom** [comment]
@zimbatm friendly ping! 😃 

Any plans for pluggable authentication? (a simple function returning a promise, set when initializing the pusher object, with some default function that you provide)

**Comment by hph** [comment]
I'm closing "feature request" issues, including this one. This does not mean that the feature will not be implemented, we just want to reserve open GitHub issues to bugs that need to be addressed.

Feel free to comment here to keep the discussion going. People watching the repository will still receive notifications.

**Comment by sandstrom** [comment]
AFAIK this was implemented (contrary to hph's message above): #211
